### PR TITLE
Cleanup delegate-related code

### PR DESCRIPTION
**Pull request by jkotas**
- Deleted unused and unreachable code
- Fix minor leak of delegate Stubs with collectible assemblies

### CODE REVIEW

File: src/coreclr/vm/class.h [hunk 1]
```diff
@@ -1895,11 +1895,9 @@ class DelegateEEClass : public EEClass
     PTR_Stub                         m_pStaticCallStub;
     PTR_Stub                         m_pInstRetBuffCallStub;
     PTR_MethodDesc                   m_pInvokeMethod;
-    PTR_Stub                         m_pMultiCastInvokeStub;
-    PTR_Stub                         m_pWrapperDelegateInvokeStub;
+    PCODE                            m_pMultiCastInvokeStub;
+    PCODE                            m_pWrapperDelegateInvokeStub;
     UMThunkMarshInfo*                m_pUMThunkMarshInfo;
-    PTR_MethodDesc                   m_pBeginInvokeMethod;
```
Comment (jkotas) [inline src/coreclr/vm/class.h:1901]: These fields were almost unused.

File: src/coreclr/vm/comdelegate.cpp [hunk 2]
```diff
@@ -2263,17 +2251,15 @@ extern "C" PCODE QCALLTYPE Delegate_GetMulticastInvokeSlow(MethodTable* pDelegat
                                                                pSig, cbSig,
                                                                NULL,
                                                                &sl);
-        pStub = Stub::NewStub(JitILStub(pStubMD));
+        pStub = JitILStub(pStubMD);
```
Comment (jkotas) [inline src/coreclr/vm/comdelegate.cpp:2254]: This is fixing memory leak. Stub allocated by Stub::NewStub would be never freed with collectible assemblies. We do not need to allocate it in the first place.

File: src/coreclr/vm/stublink.cpp [hunk 3]
```diff
@@ -350,7 +350,6 @@ StubLinker::StubLinker()
     m_pCodeElements     = NULL;
     m_pFirstCodeLabel   = NULL;
     m_pFirstLabelRef    = NULL;
-    m_pPatchLabel       = NULL;
```
Comment (jkotas) [inline src/coreclr/vm/stublink.cpp:353]: Dead code

File: src/coreclr/vm/stublink.h [hunk 4]
```diff
@@ -477,12 +468,12 @@ class Stub
         LOADER_HEAP_BIT         = 0x20000000,
         INSTANTIATING_STUB_BIT  = 0x10000000,
         UNWIND_INFO_BIT         = 0x08000000,
-        THUNK_BIT               = 0x04000000,
+        SHUFFLE_THUNK_BIT       = 0x04000000,
```
Comment (jkotas) [inline src/coreclr/vm/stublink.h:471]: Renaming to what it actually means

File: src/coreclr/vm/stubmgr.cpp [hunk 5]
```diff
@@ -1345,89 +1336,14 @@ static BOOL TraceManagedThunk(
...
-    int offsetOfNextDest = (pc == GetEEFuncEntryPoint(SinglecastDelegateInvokeStub))
-        ? DelegateObject::GetOffsetOfMethodPtr()
-        : DelegateObject::GetOffsetOfMethodPtrAux();
-    destAddr = *(PCODE*)(pThis + offsetOfNextDest);
-
-#else
-    PORTABILITY_ASSERT("TraceManagedThunk");
-    destAddr = (PCODE)NULL;
-#endif
+    *pRetAddr = (BYTE *)StubManagerHelpers::GetReturnAddress(pContext);
```
Comment (jkotas) [inline src/coreclr/vm/stubmgr.cpp:1339]: Reimplement the method using x-plat helpers and delete some unreachable code.

File: src/coreclr/vm/codeman.h [hunk 6]
```diff
@@ -108,9 +108,6 @@ enum StubCodeBlockKind : int
     STUB_CODE_BLOCK_NOCODE,
     STUB_CODE_BLOCK_MANAGED,
     STUB_CODE_BLOCK_STUBLINK,
-    // Placeholdes used by NGen images
-    STUB_CODE_BLOCK_VIRTUAL_METHOD_THUNK,
```
Comment (jkotas) [inline src/coreclr/vm/codeman.h:112]: Dead code

**Comment by jkotas** [comment]
Cleanup motivated by codereview of #104731

File: src/coreclr/vm/stubmgr.cpp [hunk 7]
```diff
@@ -1285,19 +1285,12 @@ BOOL StubLinkStubManager::DoTraceStub(PCODE stubStartAddress,
         LOG_TRACE_DESTINATION(trace, stubStartAddress, "StubLinkStubManager(InstantiatingMethod)::DoTraceStub");
         return TRUE;
     }
-    else if (stub->IsManagedThunk())
+    else if (stub->IsShuffleThunk())
     {
         trace->InitForManagerPush(stubStartAddress, this);
         LOG_TRACE_DESTINATION(trace, stubStartAddress, "StubLinkStubManager(ManagedThunk)::DoTraceStub");
```
Comment (hoyosjs) [inline src/coreclr/vm/stubmgr.cpp:1291]:
```suggestion
        LOG_TRACE_DESTINATION(trace, stubStartAddress, "StubLinkStubManager(ShuffleThunk)::DoTraceStub");
```

File: src/coreclr/vm/stubmgr.cpp [hunk 8]
```diff
@@ -1475,10 +1391,10 @@ BOOL StubLinkStubManager::TraceManager(Thread *thread,
         BYTE *pbDel = (BYTE *)StubManagerHelpers::GetThisPtr(pContext);
         return TraceDelegateObject(pbDel, trace);
     }
-    else if (stub->IsManagedThunk())
+    else if (stub->IsShuffleThunk())
     {
         LOG((LF_CORDB,LL_INFO10000, "SLSM:TM ManagedThunk\n"));
```
Comment (noahfalk) [inline src/coreclr/vm/stubmgr.cpp:1396]:
```suggestion
        LOG((LF_CORDB,LL_INFO10000, "SLSM:TM ShuffleThunk\n"));
```

File: src/coreclr/vm/stubmgr.cpp [hunk 9]
```diff
@@ -1285,19 +1285,12 @@ BOOL StubLinkStubManager::DoTraceStub(PCODE stubStartAddress,
...
     }
-    else if (stub->GetPatchOffset() != 0)
-    {
-        // The patch offset is currently only non-zero in x86 non-IL delegate scenarios.
-        trace->InitForFramePush((PCODE)stub->GetPatchAddress());
-        LOG_TRACE_DESTINATION(trace, stubStartAddress, "StubLinkStubManager::DoTraceStub");
-        return TRUE;
-    }
 
     LOG((LF_CORDB, LL_INFO10000, "StubLinkStubManager::DoTraceStub: patch offset is 0!\n"));
```
Comment (hoyosjs) [inline src/coreclr/vm/stubmgr.cpp:1295]: Add assert to make sure this is true? And add the offset to the log for debugging purposes?

File: src/coreclr/vm/stubmgr.cpp (hunk 7 above)
Comment (noahfalk) [inline src/coreclr/vm/stubmgr.cpp:1291]:
```suggestion
        LOG_TRACE_DESTINATION(trace, stubStartAddress, "StubLinkStubManager(ShuffleThunk)::DoTraceStub");
```

**Review by hoyosjs — APPROVED** [review]
The stuff in the stubmgr and such lgtm

**Review by noahfalk — APPROVED** [review]
LGTM, a few minor suggestions inline

File: src/coreclr/vm/comdelegate.cpp [hunk 10]
```diff
@@ -2147,12 +2135,12 @@ extern "C" PCODE QCALLTYPE Delegate_GetMulticastInvokeSlow(MethodTable* pDelegat
...
     _ASSERTE(pDelegateMT != NULL);
 
-    PCODE fptr = (PCODE)NULL;
+    PCODE pStub = (PCODE)NULL;
 
     BEGIN_QCALL;
 
     DelegateEEClass *delegateEEClass = (DelegateEEClass*)pDelegateMT->GetClass();
-    Stub *pStub = delegateEEClass->m_pMultiCastInvokeStub;
+    PCODE pStub = delegateEEClass->m_pMultiCastInvokeStub;
```
Comment (huoyaoyuan) [inline src/coreclr/vm/comdelegate.cpp:2143]: There are two variables named `pStub` in different scopes introduced by the QCall scope. The outer one is shadowed thus never assigned and the method would always return null.

Nit: is there a better name? `pCode` is already taken.

File: src/coreclr/vm/stubmgr.cpp (hunk 9 above)
Comment (jkotas) [reply src/coreclr/vm/stubmgr.cpp]: Patch offset does not exist after this change. I am just going to update the message.

File: src/coreclr/vm/comdelegate.cpp (hunk 10 above)
Comment (jkotas) [reply src/coreclr/vm/comdelegate.cpp:2143]: This was an oversight. There should be just one pStub variable in the method.

File: src/coreclr/vm/stublink.h [hunk 11]
```diff
@@ -834,7 +789,6 @@ class Stub
         UINT32 m_numCodeBytesAndFlags;
         union
         {
-            USHORT          PatchOffset;
             PTR_MethodDesc  InstantiatedMethod;
         } m_data;
```
[the remark this answers is not in the document]
Comment (jkotas) [reply src/coreclr/vm/stublink.h:793]: I left the union around in case we need to store some extra data for another stub type. I will add a comment.

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

**Pull request merged**